src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LspEditorFeatureDetector.cs

```diff
@@ -90,7 +109,69 @@ public bool IsLspEditorEnabled()
             _jtf.AssertUIThread();
         }
 
-        return !_lazyUseLegacyEditorTask.GetValue(_disposeTokenSource.Token);
+        _activityLog.LogInfo("Checking if LSP Editor is available");
+        if (documentMoniker is null)
+        {
+            _activityLog.LogWarning($"LSP Editor not available because {nameof(documentMoniker)} is null");
+            return false;
+        }
+        
+        var isLspEditorEnabled = !_lazyUseLegacyEditorTask.GetValue(_disposeTokenSource.Token);
+
+        if (!isLspEditorEnabled)
+        {
+            _activityLog.LogInfo("Using Legacy editor because the option or feature flag was set to true");
+            return false;
+        }
+
+        if (!ProjectSupportsLSPEditor(documentMoniker))
+        {
+            // Current project hierarchy doesn't support the LSP Razor editor
+            _activityLog.LogInfo("Using Legacy editor because the current project does not support LSP Editor");
+            return false;
+        }
+
+        return true;
+    }
+
+    // NOTE: This code is needed for legacy Razor editor support in .Net Framework projects. Do not delete unless support for .Net Framework projects is discontinued.
+    private bool ProjectSupportsLSPEditor(string documentMoniker)
+    {
+        var hr = _vsUIShellOpenDocument.Value.IsDocumentInAProject(documentMoniker, out var uiHierarchy, out _, out _, out _);
+        var hierarchy = uiHierarchy as IVsHierarchy;
+        if (!ErrorHandler.Succeeded(hr) || hierarchy is null)
+        {
+            if (!ErrorHandler.Succeeded(hr))
+            {
+                _activityLog.LogWarning($"Project does not support LSP Editor because {nameof(_vsUIShellOpenDocument.Value.IsDocumentInAProject)} failed with exit code {hr}");
+            }
+            else if (hierarchy is null)
+            {
+                _activityLog.LogWarning($"Project does not support LSP Editor because {nameof(hierarchy)} is null");
+            }
+
+            return false;
+        }
```

Comment:
Nit:
```suggestion
        if (!ErrorHandler.Succeeded(hr))
        {
            _activityLog.LogWarning($"Project does not support LSP Editor because {nameof(_vsUIShellOpenDocument.Value.IsDocumentInAProject)} failed with exit code {hr}");
            return false;
        }
        
        if (hierarchy is null)
        {
            _activityLog.LogWarning($"Project does not support LSP Editor because {nameof(hierarchy)} is null");
            return false;
        }
```